### PR TITLE
Implemented parallel::stable_partition

### PR DESCRIPTION
**Pull request by hkaiser**
This is related to #1141

### CODE REVIEW

File: hpx/parallel/algorithms/partition.hpp [hunk 1]
```diff
@@ -0,0 +1,237 @@
...
+                    std::forward<ExPolicy>(policy), first, last, size,
+                    std::forward<F>(f), std::forward<Proj>(proj),
+                    size == 1 ? 1 : (std::min)(size, max_chunks));
+            }
+        };
+        /// \endcond
+    }
+
+    /// Permutes the elements in the range [first, last) such that there exists
+    /// and iterator i such that for every iterator j in the range [first, i)
```
Comment (biddisco) [inline hpx/parallel/algorithms/partition.hpp]: 'an' not 'and'

File: tests/unit/parallel/algorithms/stable_partition_tests.hpp [hunk 2]
```diff
@@ -0,0 +1,281 @@
...
+            ++count;
+            return v1 == v2;
+        }));
+    HPX_TEST_EQ(count, d.size());
+}
+
+template <typename ExPolicy, typename IteratorTag>
+void test_stable_partition_async(ExPolicy p, IteratorTag)
+{
+//     typedef std::vector<int>::iterator base_iterator;
```
Comment (biddisco) [inline tests/unit/parallel/algorithms/stable_partition_tests.hpp]: a lot of commented out code

Comment (hkaiser) [inline tests/unit/parallel/algorithms/stable_partition_tests.hpp]: Uhh, and I was already wondering why my tests are all passing... ;-)

**Comment by hkaiser** [comment]
@biddisco All review comments have been addressed.

File: hpx/parallel/util/detail/algorithm_result.hpp [hunk 3]
```diff
@@ -31,12 +31,14 @@ namespace hpx { namespace parallel { namespace util { namespace detail
             return T();
         }
 
-        static type get(T && t)
+        template <typename T_>
+        static type get(T_ && t)
```
Comment (sithhell) [inline hpx/parallel/util/detail/algorithm_result.hpp]: What's the rationale for this change?

Comment (hkaiser) [inline hpx/parallel/util/detail/algorithm_result.hpp]: I need to revert this change and see what breaks to answer this question ;)

**Comment by sithhell** [comment]
On 10/04/2016 03:40 PM, Hartmut Kaiser wrote:
> *@hkaiser* commented on this pull request.
> 
> ------------------------------------------------------------------------
> 
> In hpx/parallel/util/detail/algorithm_result.hpp
> <https://github.com/STEllAR-GROUP/hpx/pull/2345>:
> 
>> @@ -31,12 +31,14 @@ namespace hpx { namespace parallel { namespace util { namespace detail
>              return T();
>          }
>  
> -        static type get(T && t)
> +        template <typename T_>
> +        static type get(T_ && t)
> 
> I need to revert this change and see what breaks to answer this question ;)

It looks at least strange since the other `get` calls expect rvalues ;)

File: tests/unit/parallel/algorithms/stable_partition_tests.hpp [hunk 4]
```diff
@@ -0,0 +1,285 @@
...
+    catch(hpx::exception_list const& e) {
+        caught_exception = true;
+        test::test_num_exceptions<ExPolicy, IteratorTag>::call(p, e);
+    }
+    catch(...) {
+        HPX_TEST(false);
+    }
+
+    HPX_TEST(caught_exception);
+    HPX_TEST(returned_from_algorithm);
```
Comment (biddisco) [inline tests/unit/parallel/algorithms/stable_partition_tests.hpp]: My point was that the algorithm is invoked and throws, but if the thread invoking the algorithm suspends immediately after calling it, the algorithm might throw before the `returned_from_algorithm` flag is set. This would make the test fail (once every few million runs) even though the actaul test itself was still correct

Comment (hkaiser) [inline tests/unit/parallel/algorithms/stable_partition_tests.hpp]: The idea is that even if the algorithm itself throws, the exception will never be visible on the calling thread until it calls `future::get`. So even if the calling thread is suspended right before the flag is set, the test has to succeed as any exception thrown by the algorithm will be rethrown in the context of the calling thread later only.
